### PR TITLE
MM-20524: Migrate 'components/rename_channel_modal' module to TypeScript

### PR DESCRIPTION
**Pull request by mathiasvr**
#### Summary

Migrate `components/rename_channel_modal` module and associated tests to TypeScript.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/13716
https://mattermost.atlassian.net/browse/MM-20524

-------

<details>
<summary><b>Edit:</b> CI saved the day.</summary>
Hello there!

I have been unable to figure out how to pass type checking without explicitly specifying the `intl` attribute on the `RenameChannelModal` element in the tests when using the `shallowWithIntl` function.

I assume I'm supposed to do this since e.g. [sidebar_direct_channel.test.tsx](https://github.com/mattermost/mattermost-webapp/blob/master/components/sidebar/sidebar_channel/sidebar_direct_channel/sidebar_direct_channel.test.tsx) is able to use `shallowWithIntl` without explicitly specifying the `intl` attribute.

I hope someone might be able to help, thanks!
</details>

### CODE REVIEW

**Review by Willyfrog — APPROVED** [review]
LGTM

File: components/rename_channel_modal/rename_channel_modal.test.tsx [hunk 1]
```diff
@@ -3,19 +3,53 @@
...
+import {Team} from 'mattermost-redux/types/teams';
 
 import {shallowWithIntl} from 'tests/helpers/intl-test-helper';
-import RenameChannelModal from 'components/rename_channel_modal/rename_channel_modal.jsx';
+
+import RenameChannelModal, {RenameChannelModal as RenameChannelModalClass} from './rename_channel_modal';
 
 describe('components/RenameChannelModal', () => {
-    const channel = {
+    const channel: Channel = {
```
Comment (devinbinnie) [inline components/rename_channel_modal/rename_channel_modal.test.tsx:14]: For future reference, we have a full `Channel` object mocked out in `utils/test_helper` if you wanted to use that.

File: components/rename_channel_modal/rename_channel_modal.test.tsx [hunk 2]
```diff
@@ -3,19 +3,53 @@
...
+        isCurrent: false,
+        teammate_id: '',
+        status: '',
+        fake: false,
+        group_constrained: false,
     };
 
-    const team = {
-        name: 'Fake Team', display_name: 'fake-team', type: 'O',
+    const team: Team = {
```
Comment (devinbinnie) [inline components/rename_channel_modal/rename_channel_modal.test.tsx:37]: Same with `Team`.

File: components/rename_channel_modal/rename_channel_modal.tsx [hunk 3]
```diff
@@ -34,61 +36,45 @@ const holders = defineMessages({
     },
 });
 
-export class RenameChannelModal extends React.PureComponent {
-    static propTypes = {
-
-        /**
```
Comment (devinbinnie) [inline components/rename_channel_modal/rename_channel_modal.tsx:40]: Can we keep these comments when transferring from `PropTypes` over to the type?

File: components/rename_channel_modal/rename_channel_modal.tsx [hunk 4]
```diff
@@ -34,61 +36,45 @@ const holders = defineMessages({
...
+        patchChannel: (channelId: string, patch: Channel) => Promise<{ data: Channel, error: Error }>;
     };
+}
+
+type State = {
+    displayName: string;
+    channelName: string;
+    serverError?: string;
+    urlErrors: JSX.Element[];
+    displayNameError: JSX.Element | string;
```
Comment (devinbinnie) [inline components/rename_channel_modal/rename_channel_modal.tsx:55]: There's a React provided type called `React.ReactNode` that includes both these types and any others that you can render to the DOM.

**Review by devinbinnie — CHANGES_REQUESTED** [review]
Looks good overall, just a couple comments.
Thanks for the contribution @mathiasvr!

**Comment by mathiasvr** [comment]
Thanks @Willyfrog and @devinbinnie.
I have addressed the comments, let me know if more changes are needed! 👍

**Review by devinbinnie — APPROVED** [review]
Looking good now, thanks @mathiasvr!

**Comment by saturninoabril** [comment]
@mathiasvr LGTM, tested and passed. Thank you for your contribution!

**Review by saturninoabril — APPROVED** [review]

**Pull request merged**